android/java/org/chromium/chrome/browser/app/domain/CryptoModel.java

```diff
@@ -122,88 +119,60 @@ public void resetServices(Context context, TxService mTxService, KeyringService
 
     public void init() {
         synchronized (mLock) {
-            if (mKeyringService == null || mPendingTxHelper == null) {
+            if (mKeyringService == null) {
                 return;
             }
             refreshTransactions();
-
-            // Filter out a separate list of unapproved transactions
-            mPendingTransactions =
-                    Transformations.map(mPendingTxHelper.mTransactionInfoLd, transactionInfos -> {
-                        List<TransactionInfo> pendingTransactionInfo = new ArrayList<>();
-                        for (TransactionInfo info : transactionInfos) {
-                            if (info.txStatus == TransactionStatus.UNAPPROVED) {
-                                pendingTransactionInfo.add(info);
-                            }
-                        }
-                        return pendingTransactionInfo;
-                    });
             updateCoinType();
             mNetworkModel.init();
         }
     }
 
     public void getPublicEncryptionRequest(Callback1<GetEncryptionPublicKeyRequest> onResult) {
-        synchronized (mLock) {
```

Comment:
We wrapped the code in `setupWalletModel` inside `postTask(TaskTraits.UI_DEFAULT` now, so everything will be executed in the UI thread from now on. 
The problem I saw here was calling nested synchronized methods on the same lock: because these methods were called from another method that already acquired the lock we were in muddy waters, AKA we were entering the dark world of _reentrant synchronization_.
A thread cannot acquire a lock owned by another thread, but instead it can acquire a lock that it already owns (in this case `mLock`). Allowing a thread to acquire the same lock more than once enables reentrant synchronization. 
Time to unleash my drawing skills 😁

Let's take for example `refreshTransactions`, we were in this situation:
```
init (sync on `mLock`)
|
|
| -> refreshTransactions (sync on `mLock`) // Already owned by thread, **reentrant synchronization** 
|      |
|      |
|      |
|      end refreshTransactions (acquisition count decremented)
|
|
end init (release `mLock`)
```


I would fall back on reentrant synchronization only as a last resort. Among the drawbacks: it's quite difficult to debug, and it also introduces a slight overhead for checking the lock.
I know, it's not always possible, but a solution that would simplify our lives would be to execute these calls (where possible) on the main thread using `postTask(TaskTraits.UI_DEFAULT` and introduce some checks like `@MainThread` annotation. 
Another possible solution would be to introduce a more strict check like `ThreadUtils.assertOnUiThread()`.
This is used in the Chromium source code, and it will throw exception if the execution is not on the UI thread.
Another solution would be to introduce a different lock (`mLock2`).

To conclude, I've checked all the methods where I've removed the `synchronized` keyword, they were all executed from `EncryptionKeyFragment` (UI thread) or from `init/reset` (already synchronized) so we should be safe. I've also tested the implementation and was not able to observe any issue (unfortunately was not able to reproduce the ANR in the first place).
If you are still skeptical, I can revert these changes, and we'll think of something else, no issue at all 😁